Midi 18/bag of pitches

WojciechMat opened the pull request:
I have implemented wandb to plot loss and accuracy for train and validation data.
I have also changed names in validation dict and in print() to have val_ before them. I find it more readable and it makes using wandb easier I think.

WojciechMat committed:
MIDI-15 + MIDI-18

WojciechMat committed:
wandb implementation

WojciechMat committed:
wandb implementation

WojciechMat committed:
started using hydra

WojciechMat committed:
config change

WojciechMat commented:
I am learning to use hydra and implemented it to take care of  hyperparameters and model layers (useful only when testing - training only one model in one run)

roszcz commented on config.py:
Please take a look at this discussion and let me know if your use case is different in any way: https://github.com/Nospoko/annotated-diffusion/pull/4#discussion_r1285257589

roszcz reviewed (commented):
Can we clean up this project a little? Right now I'm seeing code for two different approaches which makes it difficult to follow. 

To improve this you can separate the code from #1 - ideally close the pull request, merge, and branch from master with the new problem. The problem separation can be done with modules - what is now `main.py`, should become a separate module dedicated to the problem that it's solving, i.e. classification based on naive pitch serialization.

WojciechMat committed:
Merge branch 'main' into MIDI-18/bag-of-pitches

WojciechMat committed:
tests on chosen composers, analyzing database and number of samples

WojciechMat committed:
notes on number of samles in database

WojciechMat committed:
training models with combinations

WojciechMat committed:
changed run name for experiment

WojciechMat commented:
## Dataset info 
### Number of samples:
I printed number of samples for each composer. This will help me analyze model performance after training for each pair of composers:
'''
# samples for each of chosen composers:
    # on train:
    # composer
    # Alexander Scriabin          294272
    # Claude Debussy              310528
    # Domenico Scarlatti           67712
    # Felix Mendelssohn           358272
    # Franz Liszt                1489536
    # Franz Schubert             2096896
    # Frédéric Chopin            1508096
    # Johann Sebastian Bach       613120
    # Johannes Brahms             419328
    # Joseph Haydn                223232
    # Ludwig van Beethoven       1441152
    # Robert Schumann             897024
    # Sergei Rachmaninoff         305280
    # Wolfgang Amadeus Mozart     230528
    # on validation:
    # composer
    # Alexander Scriabin          40960
    # Felix Mendelssohn           22656
    # Franz Liszt                182144
    # Franz Schubert              17536
    # Frédéric Chopin            179200
    # Johann Sebastian Bach       56576
    # Johannes Brahms            107136
    # Joseph Haydn                29056
    # Ludwig van Beethoven       198144
    # Robert Schumann            138880
    # Sergei Rachmaninoff        116608
    # Wolfgang Amadeus Mozart     54784
    # on test:
    # composer
    # Alexander Scriabin          33024
    # Claude Debussy              68352
    # Domenico Scarlatti          39168
    # Felix Mendelssohn           16512
    # Franz Liszt                230272
    # Franz Schubert             255616
    # Frédéric Chopin            224640
    # Johann Sebastian Bach       62848
    # Johannes Brahms             26624
    # Joseph Haydn                31488
    # Ludwig van Beethoven       228480
    # Robert Schumann            112256
    # Sergei Rachmaninoff         70144
    # Wolfgang Amadeus Mozart     34688
    #
    # pieces of composers:
    # Alexander Scriabin          22 # val: None
    # Claude Debussy              37
    # Domenico Scarlatti          18
    # Felix Mendelssohn           24 #test: 1 val: 3
    # Franz Liszt                 93
    # Franz Schubert             159
    # Frédéric Chopin            145
    # Johann Sebastian Bach      114
    # Johannes Brahms             20 #test: 1 val: 5
    # Joseph Haydn                29
    # Ludwig van Beethoven       110
    # Robert Schumann             33
    # Sergei Rachmaninoff         29
    # Wolfgang Amadeus Mozart     27
'''
## Training
I started training models for each pair of the composers with at least 10 pieces in the dataset. I will evaluate each performace when the training will have ended. 

Each model's weights can be saved to a file by specifying model.save_model parameter as 'y'.
### Initial results:
For Alexander Scriabin results on a simple NN ([128, 128, 256, 64, 2] neurons) seems pretty much the same:
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/f54bf4fc-f2c3-4f2c-a50b-a63f3f01e7d7)

WojciechMat committed:
changed run name for experiment

WojciechMat committed:
update docstring for run_experiment

roszcz commented:
@WojciechMat thanks, this looks great! 

Can you try to find a configuration where the model is able to learn something? I would use two of the more popular composers with fairly distinct musical styles - like Chopin vs. Bach. Try to find a set of hyper parameters where the validation loss will actually go down. My initial suggestion would be `layers = [128, 64, 2]` and `learning_rate=1e-4` :)

roszcz commented on train.py:
```diff
@@ -0,0 +1,164 @@
+    run = wandb.init(
+        project="MIDI-18-bag-of-pitches",
+        name=f"{classnames[0].replace(' ', '_')}_vs_{classnames[1].replace(' ', '_')}-{run_id}",
+        config={
+            "learning_rate": cfg.hyperparameters.lr,
+            "n_epochs": cfg.hyperparameters.num_epochs,
+            "architecture": "NN",
+            "batch_size": cfg.hyperparameters.batch_size,
+        },
```
```suggestion
        config=OmegaConf.to_container(cfg, resolve=True),
```

We can store the full hydra config in wandb

roszcz commented on train.py:
```diff
@@ -0,0 +1,164 @@
+        print(f"{pair[0]} vs {pair[1]}")
+        model = run_experiment(cfg=cfg, train_data=train_dataloader, val_data=v_dataloader, classnames=pair)
+        models.append((model, pair))
+        path = f"models/{pair[0].replace(' ', '_').lower()}-{pair[1].replace(' ', '_').lower()}.pth"
```
```suggestion
        first_composer = pair[0].replace(' ', '_').lower()
        other_composer = pair[1].replace(' ', '_').lower()
        path = f"models/{first_composer}-{other_composer}.pth"
```

I think it's best to avoid writing code inside strings :)

roszcz commented on train.py:
```diff
@@ -0,0 +1,164 @@
+    models = []
+
+    # train model for each pair
+    for pair in composers:
```
Not sure if I'm missing something here, but composer pair is an argument in the `ComposerClassificationDataset`, and you are not passing this argument at any point - it's only used for the experiment name. 

So, it looks like this loop will always run on the default set of Chopin vs Beethoven?

WojciechMat replied:
Yeah, I realized that, I am running the training again : ))))

WojciechMat committed:
fixed bugs changed model

WojciechMat commented:
I have changed model to [128, 64, 2] and fixed bug with training on the same set each time. :’)
Now the results seem more realistic:
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/f99ee73d-97cc-4653-8a52-815473dae646)

roszcz commented:
Nice, but I think it's still suspicious that you get accuracy 1 on val dataset since epoch 1 for Scarlatti vs Scriabin :thinking: I'm guessing it's because those composers are not well represented. Can you try to focus only on top composers from this dataset until you are certain that the model is learning properly?

WojciechMat commented:
Sure, I was going to remove them after checking which are well represented and that val_accuracy does not result only from size differences.

WojciechMat committed:
training model for each pair of well-represented composers

WojciechMat committed:
automatic output neurons in PitchSeqNN, plotting matrixes for interesting pairs, utils and other code updates

WojciechMat commented:
The code is a bit of a mess, I will fix that later.
## Training results
(base_acc means max(num_samples1, num_samples2) / num_samples1 + num_samples2)

**Parameters:**
hyperparameters:
  num_epochs: 55
  batch_size: 32
  lr: 1e-4
model:
  hidden_layers: [128, 64]

I have trained a [128, 128, 64, 2] layered neural network on pairs of composers with at last 5 pieces in validation split of the dataset. Here are the most interesting results.
### Good performance
**Scriabin vs Bach**
The model reached up to 91% accuracy on validation (base_acc: 0.58, diff: 0.33):

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/26b7acf0-7215-4214-9970-030fffecf702)

Confusion matrix on test (accuracy: 0.86):

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/c9bf786d-d1b8-4a44-b2fe-47effcc6d2e8)

**Liszt vs Bach**
val_accuracy: 0.88 (base_acc: 0.76, diff: 0.22)

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/2fe4d8a1-216a-4ba1-933c-2c82ee149087)

Confusion matrix on test (accuracy: 0.90):
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/f34e35b5-40e2-4512-95a2-c5ed78f0cf7b)

### Poor performance:
**Scriabin vs Rachmaninoff**
val_accuracy: 0.59 (base_acc: 0.63)

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/761fdcbb-2d71-4165-a2a2-9a0e8ea43b36)

Confusion matrix on test (accuracy: 0.50):
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/ac983419-7fbd-4bc6-a081-f89af751af6e)

**Schumann vs Rachmaninoff**
val_accuracy: 0.58 (base: 0.54)

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/62f9e7e9-ba22-402d-b7fb-a84602d0b636)

Confusion Matrix on test (accuracy: 0.54):

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/8b60c1fa-ce75-4ce6-b289-035a9e635e74)

**Schumann vs Beethoven**
val_accuracy: 0.64 (base: 0.59)

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/6bad3734-19f6-41eb-8ce8-5ebbcf4f3606)

Confusion matrix on test (accuracy: 0.66) :

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/bb9a60a5-87f0-4aba-ab64-d29b6703f8b1)

### Plots:
Two best and three worst pairs:
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/105490e0-474a-49be-9464-13ee749bd657)

**Scriabin vs Bach vs Liszt vs Beethoven**
val_accuracy: 0.60

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/47fa18d1-f274-4885-bbc5-0c0c42bc588e)

Confusin matrix on test (accuracy: 0.66):
![Screenshot from 2023-08-09 13-30-13](https://github.com/Nospoko/bert-midi-evals/assets/74838859/3d78e88e-5065-46e4-b79a-c6789cae3781)


## Next steps:
I plan on finding a more reliable model by experimenting with Scriabin and Bach pair. I will also create a function to randomly select some samples where the model is failing and visualize them (with Streamlit maybe?)

WojciechMat committed:
update path

roszcz commented on train.py:
```diff
@@ -0,0 +1,253 @@
+    """
+    path = classnames[0].replace(" ", "_").lower()
+    for classname in classnames[1:]:
+        path += f"-{classname.replace(' ', '_').lower()}"
```
```suggestion
    path = "-".join([composer.replace(" ", "_") for composer in classnames])
```

I think this is more pythonic, and there's no code within a string :)

roszcz commented on train.py:
```diff
@@ -0,0 +1,253 @@
+    for classname in classnames[1:]:
+        path += f"-{classname.replace(' ', '_').lower()}"
+    path += ".pth"
+    if not os.path.isfile(f"models/{path}"):
+        model = run_experiment(cfg, "MIDI-18-bag-of-pitches-pairs", classnames=classnames)
+        if cfg.model.save_model == "y":
+            torch.save(model.state_dict(), f"models/{path}.pth")
+    else:
+        model = PitchSeqNN(cfg.model.hidden_layers, 128, len(classnames))
+        model.load_state_dict(torch.load(f"models/{path}.pth"))
```
I think it would be better to try and get rid of this `if` statement - if the point of this function is "test", there shouldn't be a logical path to train the model inside  of it as well.

If you want to have a single method that takes care of the full experiment for a pair of composers, I think it's better to have more generic name, like "def process_composers()", rather than precisely descriptive, like "def do_A_and_B_and_maybe_C()":

```python
def process_composers(composers: list[str], cfg: DictConfig) -> dict:
  model = train_classifier(composers, cfg)
  validation_results = validate_classifier(model, cfg)
  return validation_results
```

This may require some untangling, but I think it's worth it. In the top function you could do:

```python
scores = []
for composer_pair in composers:
  score = process_composers(composer_pair)
  score["composer_pair"] = composer_pair
  scores.append(score)
```

And use the `scores` object if you want to do any result visualization or processing. But I think this should be used more for a local review and to design better metrics to push to wandb :)

roszcz commented on train.py:
```diff
@@ -0,0 +1,253 @@
+            torch.save(model.state_dict(), path)
+
+
+def find_composers_to_check():
```
```suggestion
def find_composers_to_check() -> list[str]:
```

Please include return type hints as well :)

roszcz commented on train.py:
```diff
@@ -0,0 +1,253 @@
+        name=f"{name}-{run_id}",
+        config=OmegaConf.to_container(cfg, resolve=True),
+    )
+    return run
```
I think you don't need to return this object and track it across the experiment, you can use global methods like `wandb.log` and `wandb.finish` instead

roszcz commented on train.py:
```diff
@@ -0,0 +1,253 @@
+    return composers_to_check
+
+
+def initialize_wandb(cfg: DictConfig, project: str, classnames: list[str]):
+    # initialize experiment on WandB with unique run id
+    run_id = str(uuid.uuid1())[:8]
+    name = classnames[0].replace(" ", "_").lower()
+    for classname in classnames[1:]:
+        name += f"-{classname.replace(' ', '_').lower()}"
+    run = wandb.init(
+        project=project,
+        name=f"{name}-{run_id}",
+        config=OmegaConf.to_container(cfg, resolve=True),
+    )
```
```suggestion
def initialize_wandb(cfg: DictConfig, classnames: list[str]):
    # initialize experiment on WandB with unique run id
    run_id = str(uuid.uuid1())[:8]
    name = classnames[0].replace(" ", "_").lower()
    for classname in classnames[1:]:
        name += f"-{classname.replace(' ', '_').lower()}"
    run = wandb.init(
        project=cfg.project,
        name=f"{name}-{run_id}",
        config=OmegaConf.to_container(cfg, resolve=True),
    )
```

Try to use the config for anything that feels hardcoded :)

WojciechMat committed:
deleted unused functions, added weighted f1 metric

WojciechMat committed:
passing all the data in samples

WojciechMat commented:
For now I have implemented some of your suggestions, deleted unnecessary functions and made function to get all the samples on which the model predicted wrong. Will implement streamlit tomorrow.

WojciechMat committed:
remove weights saving

WojciechMat committed:
better f1_score calculation

roszcz commented on conf/config.yaml:
```diff
@@ -0,0 +1,11 @@
+  save_model: n
+logger:
+  run_name: "run-1"
+  project: "MIDI-18-bag-of-pitches"
```
```suggestion
train:
  num_epochs: 60
  batch_size: 16
  lr: 3e-5
  
model:
  hidden_layers: [128, 64]
  save_model: n
  
logger:
  run_name: "run-1"
  project: "MIDI-18-bag-of-pitches"
```

"train" is much better config section name than "hyperparameters" (everything in config is a hyperparameter)

roszcz commented on train.py:
```diff
@@ -0,0 +1,228 @@
+        model = run_experiment(cfg=cfg, classnames=pair)
+        models.append((model, pair))
+        
+        
```
```suggestion


```

Didn't flake8 in pre-commit hooks complain about that? 🤔

roszcz commented on train.py:
```diff
@@ -0,0 +1,228 @@
+    running_loss = running_loss / len(train_dataloader)
+    accuracy = correct / len(train_dataloader.dataset)
+    f1 = f1_score(truths, preds, average="weighted")
+    stats = {"loss": running_loss, "accuracy": accuracy, "f1_score": f1}
```
```suggestion
    stats = {"train/loss": running_loss, "train/accuracy": accuracy, "train/f1_score": f1}
```

This will create a separate "train" section in the wandb dashboar, and it's very convenient to separate metrics from different splits.

You can also add this prefix after the epoch run, like:

```python
train_stats = {f"train/{key}": value for key, value in train_stats.items()}

roszcz commented on train.py:
```diff
@@ -0,0 +1,228 @@
+    test_dataloader = DataLoader(test_data, batch_size=cfg.hyperparameters.batch_size, shuffle=True)
+
+    # test model and get all wrongly predicted samples
+    wrongly_predicted = wrong_preds(model, test_dataloader)
```
```suggestion
    bad_predictions = wrong_preds(model, test_dataloader)
```

Using adverbs for variable names should be avoided. When naming variables, you typically want nouns or adjectives because variables represent entities or states, not actions or qualities of actions.

roszcz commented on utils.py:
```diff
@@ -111,30 +116,78 @@ def make_confusion_matrix(y_true, y_pred, classes=None, figsize=(10, 10), text_s
-        for data in test_data:
-            notes, labels = data
-            out = model(notes)
+        for samples in test_data:
```
```suggestion
        for batch in test_data:
```

It's good to be consistent with the variable names, and when iterating over a `DataLoader`, we get batches

roszcz commented on utils.py:
```diff
@@ -111,30 +116,78 @@ def make_confusion_matrix(y_true, y_pred, classes=None, figsize=(10, 10), text_s
     """
-    Test the performance of the trained ComposerClassifier model with the provided test data loader
-    .by plotting a confusion matrix.
+    Test the performance of the trained ComposerClassifier model with the provided test data loader.
```
```suggestion
    Creates a shuffled copy of the input dataset in as a dict, with additional 'pred' column for model predictions.
```

I think this would be a more accurate description of the function :)

This goes beyond this function, but here's an approach that I think would be better:

```python
test_loader = DataLoader(test_dataset, shuffle=False)
predictions = []
for batch in test_loader:
  out = model(batch["data"])
  batch_prediction = out.argmax(1)
  predictions.append(batch_prediction)

predictions = torch.hstack(predictions)
target = [s["label"] for s in test_dataset]
results = predictions == torch.tensor(target)
```

Here `results` is a boolean array that can be used as indices to the dataset, e.g.:

```python
# Review bad predictions
idxs = np.where(~results)[0]
samples = [test_dataset[idx] for idx in idxs]
...

# Review per-piece performance
df = pd.DataFrame([s["title"] + "_" + s["composer"] for s in test_dataset], columns=["piece"])
df["result"] = results
groups = df.groupby("piece")

for piece, gdf in groups:
  print(piece, gdf.result.mean())
```

That way you can see the performance of your model on a piece-by-piece basis and find outliers:

```
Piece: Sonata No. 3 in C Major, Op. 2 No. 3, First Movement_Ludwig van Beethoven
Accuracy: 0.94
---
Piece: Transcendental Etude No. 3 "Paysage"_Franz Liszt
Accuracy: 0.34
```

WojciechMat replied:
okay, will do

roszcz commented on data/dataset.py:
```diff
@@ -51,3 +52,85 @@ def __len__(self) -> int:
+
+        return sample
+
+    # samples for each of chosen composers on train:
+    # composer
+    # Alexander Scriabin          294272
+    # Claude Debussy              310528
```
I think we can remove this whole comment block - you can add code to generate those stats and show how to use it in the readme, but I think we don't need access to this whole table :)

roszcz reviewed (commented):
Thanks, this is great!

The one important missing element is model checkpoints - the official [tutorial](https://pytorch.org/tutorials/recipes/recipes/saving_and_loading_a_general_checkpoint.html#save-the-general-checkpoint) is a good starting point, you can also check what we're doing in other projects.

Other than that, can you also update the README file with information about how to run this, and what the expected results are?

WojciechMat committed:
quick fixes

WojciechMat committed:
streamlit demo

roszcz commented on train.py:
```diff
@@ -168,7 +189,7 @@ def run_experiment(cfg: DictConfig, classnames: list[str]) -> nn.Module:
             "model_state_dict": model.state_dict(),
             "optimizer_state_dict": optimizer.state_dict(),
-            "loss": cfg.train.loss,
+            "lr": cfg.train.lr,
```
```suggestion
            "config": OmegaConf.to_container(cfg, resolve=True),
```

It's good practice to save the full config.

WojciechMat committed:
more plots in one run

WojciechMat commented:
I have implemented streamlit with my previous approach. To run it you can use 
streamlit run --server.port 4466 train.py

roszcz commented on train.py:
```diff
@@ -0,0 +1,274 @@
+def main(cfg: DictConfig):
+    classnames = ["Ludwig van Beethoven", "Franz Liszt"]
+    path = "models/ludwig-van-beethoven-franz-liszt.pt"
+    if os.path.isfile(path):
+        checkpoint = torch.load(path)
+        model = PitchSeqNN(cfg.model.hidden_layers, 128, 2)
+        model.load_state_dict(checkpoint["model_state_dict"])
+    else:
+        model = run_experiment(cfg, classnames=classnames)
```
I think it's better to avoid mixes of model training and model loading like that. You can think about model "evaluation" as a completely separate process, that you only run after training is finished.

When using hydra, we can have separate configs for training and evaluation. Separate entry points can also be convenient:

```sh
# This will create and print a run id ...
python train.py composers=["bach", "chopin"]

# ... which we can use to run evaluations
python eval.py run_id="run id"
```

Regarding streamlit: I still don't have any best practices or usage patterns figured out for it, but my intuition is to also treat it as a separate procedure, that can use the `eval` module.

```py
# dashboard.py
import streamlit as st

from eval import some_eval

# Maybe st.combobox wiht available run ids?
checkpoint = load_checkpoint()
...
```
And then:
```sh
streamlit dashboard.py
```

With this approach you have 3 separate sub-applications to manage, and it should be easier to keep the logic clean and separated :)

WojciechMat replied:
I was just thinking of doing that today! thanks!

WojciechMat committed:
whole config

Co-authored-by: Tomek Roszczynialski <roszcz@users.noreply.github.com>

WojciechMat committed:
eval program

WojciechMat committed:
cosmetics

WojciechMat committed:
dashboard

WojciechMat committed:
update readme

WojciechMat committed:
predicted label on streamlit

WojciechMat commented:
I have divided the project into three programs: train.py, eval.py and dashboard.py and initially updated readme.

roszcz commented on train.py:
```diff
@@ -0,0 +1,221 @@
+        path += f"{classname.lower().replace(' ', '_')}-"
+    path += f"{run_id}.pt"
+
+    torch.save(
+        {
+            "epoch": cfg.train.num_epochs,
+            "model_state_dict": model.state_dict(),
+            "optimizer_state_dict": optimizer.state_dict(),
+            "config": OmegaConf.to_container(cfg, resolve=True),
+        },
+        path,
+    )
+    return model, run_id
```
I'm getting `RuntimeError: Parent directory models does not exist.` here - you can add code that will create all required directories before running the experiment, or add them to the repository by using some kind of `.gitkeep` approach

roszcz commented on README.md:
````diff
@@ -2,6 +2,27 @@
+## Dashboard
+You can look through samples on which model performs good/bad. To start, run:
+```shell
+streamlit run --server.port 4466 dashboard.py
````
I think `streamlit` is missing in requirements.txt :)

roszcz commented on utils.py:
```diff
@@ -111,30 +116,77 @@ def make_confusion_matrix(y_true, y_pred, classes=None, figsize=(10, 10), text_s
+    # stolen from Tomek
+    midi_file = os.path.basename(piece.source["midi_filename"])
+    mp3_path = midi_file.replace(".midi", ".mp3")
+    mp3_path = os.path.join("tmp", mp3_path)
```
Make sure this folder exists - for "tmp" I'd say it's the best to include it in the repo by commiting `tmp/.keep` file and addint `tmp/` to .gitignoe.

roszcz reviewed (approved):
🚀

WojciechMat committed:
tmp and models directories

WojciechMat committed:
requirements update

WojciechMat committed:
full version of dashboard

WojciechMat committed:
comment

WojciechMat commented:
## Dashboard
### Samples
You can now look through right and wrong predictions from test dataset. 
<img src="https://github.com/Nospoko/bert-midi-evals/assets/74838859/c023c308-70e6-4d31-beb9-4df07f4735aa"><img src="https://github.com/Nospoko/bert-midi-evals/assets/74838859/b4d4c610-2e46-4f58-a382-35a5ba2a06ca">

### Accuracy per piece
You can also look through all the pieces and model accuracy on each of them.
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/d2c601fe-9f27-49c7-9fd2-cc40e50b4a9b)
